tinker_cookbook/tests/test_renderers.py

```diff
@@ -373,6 +539,82 @@ def test_supervised_example_against_hf_chat_templates(model_name: str, conv_id:
     )
 
 
+# =============================================================================
+# Tokenization Boundary Difference Tests
+# =============================================================================
+
+
+def _get_conversation_with_whitespace_boundary() -> list[Message]:
+    """Conversation with leading whitespace in TextPart - triggers tokenization boundary issue.
+
+    When content parts start with whitespace (like '\\n\\n'), our renderer tokenizes
+    piece-by-piece (header, then content) while HF's apply_chat_template tokenizes
+    the full string at once. This produces different token sequences that decode
+    to identical strings.
+
+    Example with Qwen3:
+    - Piece-wise: encode("<|im_start|>assistant\\n") + encode("\\n\\nasst_text")
+      → [..., 198, 271, ...] (newline + double-newline as separate tokens)
+    - All-at-once: encode("<|im_start|>assistant\\n\\n\\nasst_text")
+      → [..., 1406, ...] (triple-newline as single token)
+
+    Both decode to the same string, but the token sequences differ.
+    """
+    return [
+        {"role": "user", "content": "Hello, how are you?"},
+        {
+            "role": "assistant",
+            "content": [
+                ThinkingPart(type="thinking", thinking="\nLet me respond politely.\n"),
+                TextPart(type="text", text="\n\nI'm fine, thank you!"),
+            ],
+        },
+    ]
+
+
+@pytest.mark.parametrize(
+    "model_name",
+    [
+        "Qwen/Qwen3-30B-A3B",  # Uses <|im_start|>assistant\n header
+    ],
+)
+def test_tokenization_boundary_with_whitespace(model_name: str):
```

Comment:
ideally remove this if we can just test the whitespace case along with the other convos